code/missionui/redalert.cpp

```diff
@@ -788,11 +791,33 @@ void red_alert_delete_ship(ship *shipp, int ship_state)
 		}
 	}
 
-	ship_add_exited_ship( shipp, Ship::Exit_Flags::Player_deleted );
+	// If this ship is docked with anybody, undock them all before deleting
+	dock_undock_all(&Objects[shipp->objnum]);
+	dock_dead_undock_all(&Objects[shipp->objnum]);
+
+	int cleanup_mode;	// See ship.h for cleanup mode defines (SHIP_VANISHED)
+	switch (ship_state) {
+	case RED_ALERT_DESTROYED_SHIP_CLASS:
+		cleanup_mode = SHIP_DESTROYED;
+		break;
+	case RED_ALERT_PLAYER_DEL_SHIP_CLASS:
+		cleanup_mode = SHIP_DEPARTED;
+		break;
+	default:
+		cleanup_mode = SHIP_VANISHED;
+		break;
+	}
+
+	Objects[shipp->objnum].flags.set(Object::Object_Flags::Should_be_dead);
+	ship_cleanup(shipnum, cleanup_mode);
+	/*
+	ship_add_exited_ship( shipp, exit_flag );
```

Comment:
`ship_cleanup` probably calls `ship_add_exited_ship` so this call isn't needed anymore?

Also, once you're done with your changes this code should be removed since commented code blocks are rarely helpful.